File: dsa-python/solutions/trees/medium/construct_binary_tree_from_preorder_and_inorder_traversal.py
# 105. Construct Binary Tree from Preorder and Inorder Traversal
# https://leetcode.com/problems/construct-binary-tree-from-preorder-and-inorder-traversal/
#
# Given two integer arrays preorder and inorder where preorder is the preorder
# traversal of a binary tree and inorder is the inorder traversal of the same
# tree, construct and return the binary tree.
#
# APPROACH: Preorder's first element = root. Find root in inorder (using HashMap
# for O(1) lookup) — everything left = left subtree, everything right = right subtree.
# Recurse. preIndex tracks current position in preorder (advances with each node created).
#
# BRUTE → OPTIMIZED: Brute scans inorder with for loop O(n) per node = O(n²).
# Optimized uses HashMap for O(1) lookup = O(n) total.
#
# TIME: O(n) — each node created once, O(1) lookup
# SPACE: O(n) — HashMap + recursion stack


class TreeNode:
    def __init__(self, val: int = 0, left: "TreeNode | None" = None, right: "TreeNode | None" = None):
        self.val = val
        self.left = left
        self.right = right


class ConstructBinaryTreeFromPreorderAndInorderTraversal:
    def build_tree(self, preorder: list[int], inorder: list[int]) -> TreeNode | None:
        inorder_map = {}
        for i in range(len(inorder)):
            inorder_map[inorder[i]] = i

        pre_index = 0

        def build(left, right):
            nonlocal pre_index
            if left > right:
                return None

            root_val = preorder[pre_index]
            pre_index += 1

            root = TreeNode(root_val)
            mid = inorder_map[root_val]

            root.left = build(left, mid - 1)
            root.right = build(mid + 1, right)

            return root

        return build(0, len(inorder) - 1)

File: dsa-csharp/solutions/Trees/Medium/ConstructBinaryTreeFromPreorderAndInorderTraversal.cs
// 105. Construct Binary Tree from Preorder and Inorder Traversal
// https://leetcode.com/problems/construct-binary-tree-from-preorder-and-inorder-traversal/
//
// Given two integer arrays preorder and inorder where preorder is the preorder
// traversal of a binary tree and inorder is the inorder traversal of the same
// tree, construct and return the binary tree.
//
// APPROACH: Preorder's first element = root. Find root in inorder using HashMap (O(1) lookup).
//           Everything left = left subtree, right = right subtree. Recurse with preIndex tracking position.
// TIME: O(n)
// SPACE: O(n)

public class ConstructBinaryTreeFromPreorderAndInorderTraversal
{
    public TreeNode? BuildTree(int[] preorder, int[] inorder)
    {
        Dictionary<int,int> inorderMap = new();

        for(int i = 0; i < inorder.Length; i++)
        {
            inorderMap[inorder[i]] = i;
        }

        int preIndex = 0;

        return build(preorder, inorderMap, 0, inorder.Length - 1, ref preIndex);
    }

    private TreeNode? build(int [] preorder, Dictionary<int,int> inorderMap, int left, int right, ref int preIndex)
    {
        if(left > right) return null;

        int rootValue = preorder[preIndex++];

        TreeNode root = new TreeNode(rootValue);

        int midValue = inorderMap[rootValue];

        root.left = build(preorder, inorderMap, left, midValue - 1, ref preIndex);
        root.right = build(preorder, inorderMap, midValue + 1, right, ref preIndex);

        return root;
    }
}
